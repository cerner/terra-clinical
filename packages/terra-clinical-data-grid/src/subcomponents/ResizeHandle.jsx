import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import { DraggableCore } from 'react-draggable';

import styles from './ResizeHandle.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * String identifer for the ResizeHandle. This value will be provided to the `onResizeStop` function prop as a parameter for
   * identification purposes.
   */
  id: PropTypes.string.isRequired,
  /**
   * Function called upon release of the ResizeHandle. The id, as well as the new ResizeHandle position, will be provided
   * as arguments. Parameters: `onResizeStop(resizeHandleId, positionDelta)`
   */
  onResizeStop: PropTypes.func,
};

class ResizeHandle extends React.Component {
  /**
   * Click events that occur on the ResizeHandle are not propagated. This prevents clicks on the ResizeHandle from registering
   * as clicks on the elements over which the ResizeHandle is rendered.
   */
  static preventClickEvent(event) {
    event.stopPropagation();
  }

  constructor(props) {
    super(props);

    this.handleDragMove = this.handleDragMove.bind(this);
    this.handleDragStart = this.handleDragStart.bind(this);
    this.handleDragStop = this.handleDragStop.bind(this);
  }

  handleDragMove(event, data) {
    const handleNode = data.node;

    this.resizeHandleDragPosition += data.deltaX;

    handleNode.style.transform = `translate3d(${this.resizeHandleDragPosition}px, 0, 0)`;
  }

  handleDragStart(event, data) {
    const handleNode = data.node;

    this.resizeHandleDragPosition = 0;

    handleNode.classList.add(cx('dragging'));
  }

  handleDragStop(event, data) {
    const { id, onResizeStop } = this.props;
    const handleNode = data.node;

    handleNode.classList.remove(cx('dragging'));
    handleNode.style.transform = '';

    if (onResizeStop) {
      onResizeStop(id, this.resizeHandleDragPosition);
    }
  }

  render() {
    return (
      /* eslint-disable jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events */
      <DraggableCore
        onStart={this.handleDragStart}
        onStop={this.handleDragStop}
        onDrag={this.handleDragMove}
      >
        <div className={cx('resize-handle')} onClick={ResizeHandle.preventClickEvent} />
      </DraggableCore>
      /* eslint-enable jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events */
    );
  }
}

ResizeHandle.propTypes = propTypes;

export default ResizeHandle;
