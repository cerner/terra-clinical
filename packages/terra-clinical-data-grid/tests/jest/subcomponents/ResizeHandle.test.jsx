import React from 'react';
import ResizeHandle from '../../../src/subcomponents/ResizeHandle';

describe('ResizeHandle Snapshot Tests', () => {
  it('should render a ResizeHandle with minimal props', () => {
    const resizeHandle = shallow((
      <ResizeHandle id="test-handle" />
    ));

    expect(resizeHandle).toMatchSnapshot();
  });

  it('should render a ResizeHandle with all props', () => {
    const resizeHandle = shallow((
      <ResizeHandle
        id="test-handle"
        onResizeStop={() => {}}
      />
    ));

    expect(resizeHandle).toMatchSnapshot();
  });
});
