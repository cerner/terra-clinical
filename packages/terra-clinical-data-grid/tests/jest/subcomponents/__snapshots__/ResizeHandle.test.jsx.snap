// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResizeHandle Snapshot Tests should render a ResizeHandle with all props 1`] = `
<DraggableCore
  allowAnyClick={false}
  cancel={null}
  disabled={false}
  enableUserSelectHack={true}
  grid={null}
  handle={null}
  offsetParent={null}
  onDrag={[Function]}
  onMouseDown={[Function]}
  onStart={[Function]}
  onStop={[Function]}
  transform={null}
>
  <div
    className="resize-handle"
    onClick={[Function]}
  />
</DraggableCore>
`;

exports[`ResizeHandle Snapshot Tests should render a ResizeHandle with minimal props 1`] = `
<DraggableCore
  allowAnyClick={false}
  cancel={null}
  disabled={false}
  enableUserSelectHack={true}
  grid={null}
  handle={null}
  offsetParent={null}
  onDrag={[Function]}
  onMouseDown={[Function]}
  onStart={[Function]}
  onStop={[Function]}
  transform={null}
>
  <div
    className="resize-handle"
    onClick={[Function]}
  />
</DraggableCore>
`;
